File: ios/FilmSimCore/Sources/FilmSimCore/FLog2.swift
import Foundation

/// F-Log2 transfer function. Constants from the F-Log2 Data Sheet Ver.1.1.
/// Reference: 18% grey -> 0.391 (10bit 400), 90% -> 0.557 (570).
/// CPU reference implementation; the GPU path is the `flog2Encode` CIColorKernel in the app target.
public enum FLog2 {
    public static let a = 5.555556
    public static let b = 0.064829
    public static let c = 0.245281
    public static let d = 0.384316
    public static let e = 8.799461
    public static let f = 0.092864
    public static let cut1 = 0.000889
    public static let cut2 = 0.100686685370811

    public static func encode(_ x: Double) -> Double {
        x >= cut1 ? c * log10(a * x + b) + d : e * x + f
    }

    public static func decode(_ y: Double) -> Double {
        y >= cut2 ? pow(10, (y - d) / c) / a - b / a : (y - f) / e
    }
}

File: research/filmsim/flog2.py
"""F-Log and F-Log2 transfer functions.

Constants are taken verbatim from the Fujifilm data sheets:
- F-Log2 Data Sheet Ver.1.1  https://dl.fujifilm-x.com/technical-data/F-Log2_DataSheet_E_Ver.1.1.pdf
- F-Log  Data Sheet Ver.1.2  https://dl.fujifilm-x.com/technical-data/F-Log_DataSheet_E_Ver.1.2.pdf

Reference points (F-Log2): 0% -> 0.093 (10bit 95), 18% -> 0.391 (400), 90% -> 0.557 (570).
"""

from __future__ import annotations

from dataclasses import dataclass

import numpy as np


@dataclass(frozen=True)
class LogCurve:
    a: float
    b: float
    c: float
    d: float
    e: float
    f: float
    cut1: float
    cut2: float

    def encode(self, x: np.ndarray | float) -> np.ndarray:
        """Scene-linear reflection (0.18 = 18% grey) -> log code value in [0, 1]."""
        x = np.asarray(x, dtype=np.float64)
        safe = np.maximum(self.a * x + self.b, 1e-12)
        log_part = self.c * np.log10(safe) + self.d
        lin_part = self.e * x + self.f
        return np.where(x >= self.cut1, log_part, lin_part)

    def decode(self, y: np.ndarray | float) -> np.ndarray:
        """Log code value in [0, 1] -> scene-linear reflection."""
        y = np.asarray(y, dtype=np.float64)
        log_part = (np.power(10.0, (y - self.d) / self.c)) / self.a - self.b / self.a
        lin_part = (y - self.f) / self.e
        return np.where(y >= self.cut2, log_part, lin_part)


FLOG2 = LogCurve(
    a=5.555556, b=0.064829, c=0.245281, d=0.384316,
    e=8.799461, f=0.092864,
    cut1=0.000889, cut2=0.100686685370811,
)

FLOG = LogCurve(
    a=0.555556, b=0.009468, c=0.344676, d=0.790453,
    e=8.735631, f=0.092864,
    cut1=0.00089, cut2=0.100537775223865,
)
